### PR TITLE
Large Langage Model for Genomics using Python

### PR DESCRIPTION
**Pull request by bebatut**
This PR:
- Creates 5 tutorials on LLM for Genomics using Python based on material from @raphaelmourad (https://github.com/raphaelmourad/LLM-for-genomics-training)
    The last tutorial is not finished, as I could not access the Google Colab

- Adds subtopic structure to statistics topic and proposes 3 topics (machine-learning, deep-learning, GAI / LLM) to organize the numerous tutorials. @anuprulez would that be okay for you?

It would be nice if this PR could be merged quickly (the tutorials are all in draft mode so they will not be displayed in the statistic page at the beginning) so the other trainers for our ELIXIR workshop in May could test them. Thanks a lot!

### CODE REVIEW

File: topics/statistics/tutorials/genomic-llm-zeroshot-prediction/tutorial.md [hunk 1]
```diff
@@ -0,0 +1,608 @@
...
+> > 1. `use_fast=True`: Enables the use of a fast tokenizer implementation, which is optimized for speed and efficiency. This is particularly useful when working with large datasets or when performance is a priority.
+> > 2. `trust_remote_code=True`: Allows the tokenizer to execute custom code from the model repository. This may be necessary for certain architectures or preprocessing steps that require additional functionality.
+> >
+> {: .solution}
+>
+{: .question}
+
+# Load and Configure the Pre-trained Model
+
+We will now load the pre-trained large language model (LLM) and configure it for our specific task of predicting the impact of DNA mutations.
```
Comment (anuprulez) [inline topics/statistics/tutorials/genomic-llm-zeroshot-prediction/tutorial.md:183]:
```suggestion
We will now load the pre-trained DNA large language model (LLM) and configure it for our specific task of predicting the impact of DNA mutations. 
```

File: topics/statistics/tutorials/genomic-llm-zeroshot-prediction/tutorial.md [hunk 2]
```diff
@@ -0,0 +1,608 @@
...
+Let's compute the hidden states for all DNA sequences:
+
+```python
+tokenized_dna = tokenizer(
+  [dna_wt, dna_mut, dna_del],
+  return_tensors="pt",
+  padding=True,
+)
+inputs_seqs = tokenized_dna["input_ids"]
+model_outputs = model(inputs)
```
Comment (anuprulez) [inline topics/statistics/tutorials/genomic-llm-zeroshot-prediction/tutorial.md:351]:
```suggestion
model_outputs = model(inputs_seqs)
```

File: topics/statistics/tutorials/genomic-llm-zeroshot-prediction/tutorial.md [hunk 3]
```diff
@@ -0,0 +1,608 @@
...
+  ```python
+  def computeEmbedding(seqs):
+    tokenized_dna = tokenizer(
+      seqs,
+      return_tensors="pt",
+      padding=True,
+    )
+    inputs_seqs = tokenized_dna["input_ids"]
+    hidden_states = model(inputs_seqs)[0].detach().cpu().numpy() 
+    return torch.max(hidden_states_seqs, dim=1)
```
Comment (anuprulez) [inline topics/statistics/tutorials/genomic-llm-zeroshot-prediction/tutorial.md:509]:
```suggestion
    return torch.max(hidden_states, dim=1)
```

File: topics/statistics/tutorials/genomic-llm-zeroshot-prediction/tutorial.md [hunk 4]
```diff
@@ -0,0 +1,608 @@
...
+> > 3. The shape is 4,096, the number of possible tokens
+> > 
+> > 4. It represents the average embedding of the DNA sequence. This fixed-size representation can be used for various downstream tasks, such as classification, clustering, or similarity comparisons.
+> {: .solution}
+>
+{: .question}
+
+# Compare the effect of mutations with or without amino acid modification
+
+Let's look at the portion of the Cystic fibrosis transmembrane conductance regulator (CFTСR) gene where a mutation responsible of the Cystic fibrosis appears.
```
Comment (anuprulez) [inline topics/statistics/tutorials/genomic-llm-zeroshot-prediction/tutorial.md:322]:
```suggestion
Let's look at the portion of the Cystic fibrosis transmembrane conductance regulator (CFTR) gene where a mutation responsible of the Cystic fibrosis appears.
```

File: topics/statistics/tutorials/genomic-llm-sequence-generation/tutorial.md [hunk 5]
```diff
@@ -0,0 +1,771 @@
...
+    os.environ["PYTORCH_CUDA_ALLOC_CONF"] = "max_split_size_mb:32"
+    ```
+
+# Generate Synthetic DNA Sequences
+
+Using the pre-trained language model which was trained on 1,011 *Saccharomyces cerevisiae* isolates, we would like to generate 100 synthetic yeast DNA sequences.
+
+## Build the Sequence Generator
+
+First, we need set up the sequence generation pipeline. This pipeline will enable us to generate synthetic DNA sequences that mimic natural genomic sequences. By leveraging the power of language models, we can create novel DNA sequences for various applications in synthetic biology.
```
Comment (anuprulez) [inline topics/statistics/tutorials/genomic-llm-sequence-generation/tutorial.md:153]:
```suggestion
First, we need to set up the sequence generation pipeline. This pipeline will enable us to generate synthetic DNA sequences that mimic natural genomic sequences. By leveraging the power of language models, we can create novel DNA sequences for various applications in synthetic biology.
```

File: topics/statistics/tutorials/genomic-llm-sequence-generation/tutorial.md [hunk 6]
```diff
@@ -0,0 +1,771 @@
...
+> > 2. 100 x 256 ($$256 = 4^{4}$$)
+> >
+> {: .solution}
+>
+{: .question}
+
+To visualize and compare real and generated DNA sequences in a continuous space, we can utilize Principal Component Analysis (PCA). PCA is a powerful dimensionality reduction technique that transforms high-dimensional data into a lower-dimensional space while preserving as much variability as possible. This allows us to visualize the similarities and differences between the sequences effectively.
+
+```python
+def plot_pca_projection(set1: np.ndarray, set2: np.ndarray, title: str = "PCA Projection") -> None:
```
Comment (anuprulez) [inline topics/statistics/tutorials/genomic-llm-sequence-generation/tutorial.md:434]: Minor suggestion: Instead of PCA for showing projections, UMAP or tSNE could also be used.

File: topics/statistics/tutorials/genomic-llm-pretraining/tutorial.md [hunk 7]
```diff
@@ -0,0 +1,1030 @@
...
+    )
+    (norm): MixtralRMSNorm((256,), eps=1e-05)
+  )
+  (lm_head): Linear(in_features=256, out_features=4096, bias=False)
+)
+```
+
+As expected, the model is a `MixtralForCausalLM` model with several key components:
+
+1. **Embedding Layer (`embed_tokens`)**: Converts input DNA sequences into dense vectors of fixed size. It maps each of the 4,096 possible DNA tokens (representing 6-mers) to a 256-dimensional vector space. This embedding layer is crucial for transforming discrete DNA sequences into a format suitable for neural network processing.
```
Comment (anuprulez) [inline topics/statistics/tutorials/genomic-llm-pretraining/tutorial.md:397]:
```suggestion
1. **Embedding Layer (`embed_tokens`)**: Converts input DNA sequences into dense vectors of fixed size. It maps each of the 4,096 (4**6) possible DNA tokens (representing 6-mers) to a 256-dimensional vector space. This embedding layer is crucial for transforming discrete DNA sequences into a format suitable for neural network processing.
```

File: topics/statistics/tutorials/genomic-llm-pretraining/tutorial.md [hunk 8]
```diff
@@ -0,0 +1,1030 @@
...
+> > 
+> > 1. 8 transformer layers
+> > 2. Yes
+> >
+> {: .solution}
+>
+{: .question}
+
+
+# Computer the embedding of a DNA sequence
```
Comment (anuprulez) [inline topics/statistics/tutorials/genomic-llm-pretraining/tutorial.md:941]:
```suggestion
# Compute the embedding of a DNA sequence 
```

File: topics/statistics/tutorials/genomic-llm-pretraining/tutorial.md [hunk 9]
```diff
@@ -0,0 +1,1030 @@
...
+
+## Install dependencies
+
+The first step is to install the required dependencies:
+
+```python
+!pip install accelerate
+!pip install datasets==3.0.1
+!pip install transformers
+!pip install torch
```
[the remark this answers is not in the document]
Comment (anuprulez) [reply topics/statistics/tutorials/genomic-llm-pretraining/tutorial.md:103]: sure, how do you execute the Python snippets? on Galaxy infrastructure or on some external machines?

[the remark this answers is not in the document]
Comment (bebatut) [reply topics/statistics/tutorials/genomic-llm-pretraining/tutorial.md:103]: We plan to use the French bioinformatics cluster with Jupyter lab

File: topics/statistics/tutorials/genomic-llm-finetuning/tutorial.md [hunk 10]
```diff
@@ -0,0 +1,725 @@
...
+notebook:
+  language: python
+  pyolite: true
+---
+
+After preparing, training, and utilizing a language model for DNA sequences, we can now fine-tune a pre-trained Large Language Model (LLM) for specific DNA sequence classification tasks. Here, we will use a pre-trained model from Hugging Face, specifically the [Mistral-DNA-v1-17M-hg38](https://huggingface.co/RaphaelMourad/Mistral-DNA-v1-17M-hg38), and adapt it to classify DNA sequences based on their biological functions. Our objective is to classify sequences according to whether they bind to transcription factors.
+
+> <comment-title>Transcription factors</comment-title>
+>
+> Transcription factors are proteins that play a crucial role in regulating gene expression by binding to specific DNA sequences, known as enhancers or promoters, near the genes they control. These proteins act as molecular switches, turning genes on or off in response to various cellular signals and environmental cues. By binding to DNA, transcription factors either promote or inhibit the recruitment of RNA polymerase, the enzyme responsible for transcribing DNA into RNA, thereby influencing the rate of transcription.
```
Comment (olisand) [inline topics/statistics/tutorials/genomic-llm-finetuning/tutorial.md:52]: Enhancers are not necessarily near the genes they control

File: topics/statistics/tutorials/genomic-llm-sequence-generation/tutorial.md [hunk 11]
```diff
@@ -0,0 +1,771 @@
...
+> >     var = np.mean(np.var(art_seq_kmer_counts, axis=0))
+> >     print(f"Variance for temperature of { temp }: { var }")
+> > ```
+> {: .solution}
+>
+{: .hands-on}
+
+> <question-title></question-title>
+>
+> 1. How similar look like the 5 first generated DNA sequences given the temparature?
```
Comment (olisand) [inline topics/statistics/tutorials/genomic-llm-sequence-generation/tutorial.md:552]:
```suggestion
> 1. How similar do the 5 first generated DNA sequences look like given the temperature?
```

File: topics/statistics/tutorials/genomic-llm-sequence-generation/tutorial.md [hunk 12]
```diff
@@ -0,0 +1,771 @@
...
+> > art_sequences = generate_synthetic_dna(model_name, max_length,  num_sequences, 0.01)
+> > art_sequences[0]
+> > ```
+> {: .solution}
+>
+{: .hands-on}
+
+> <question-title></question-title>
+>
+> 1. How many significant similarity have been found?
```
Comment (olisand) [inline topics/statistics/tutorials/genomic-llm-sequence-generation/tutorial.md:597]:
```suggestion
> 1. How many significant similarities have been found?
```

File: topics/statistics/tutorials/genomic-llm-sequence-generation/tutorial.md [hunk 13]
```diff
@@ -0,0 +1,771 @@
...
+> > art_sequences = generate_synthetic_dna(model_name, max_length,  num_sequences, 1)
+> > art_sequences[0]
+> > ```
+> {: .solution}
+>
+{: .hands-on}
+
+> <question-title></question-title>
+>
+> 1. How many significant similarity have been found?
```
Comment (olisand) [inline topics/statistics/tutorials/genomic-llm-sequence-generation/tutorial.md:630]:
```suggestion
> 1. How many significant similarities have been found?
```

**Review by olisand — COMMENTED** [review]
It was mostly a first PR review exercise.

File: topics/statistics/tutorials/genomic-llm-finetuning/tutorial.md (hunk 10 above)
Comment (bebatut) [inline topics/statistics/tutorials/genomic-llm-finetuning/tutorial.md:52]:
```suggestion
> Transcription factors are proteins that play a crucial role in regulating gene expression by binding to specific DNA sequences, known as enhancers or promoters. These proteins act as molecular switches, turning genes on or off in response to various cellular signals and environmental cues. By binding to DNA, transcription factors either promote or inhibit the recruitment of RNA polymerase, the enzyme responsible for transcribing DNA into RNA, thereby influencing the rate of transcription.
```

File: topics/statistics/tutorials/genomic-llm-sequence-generation/tutorial.md [hunk 14]
```diff
@@ -0,0 +1,771 @@
...
+```
+
+After translating a synthetic DNA sequence into a protein sequence, the next crucial step is to analyze this sequence to uncover its potential biological function. This involves identifying known protein domains or motifs within the sequence that may indicate specific functions, such as binding sites or active regions. Verifying the biological plausibility of the translated sequence is essential, which can be achieved by comparing it to known protein sequences in databases like UniProt or GenBank using tools like BLAST. Additionally, examining the conservation of the sequence across different species can provide insights into its functional importance. If the sequence shows potential for biological function, it can serve as a starting point for further research, including structural modeling to predict its three-dimensional structure, functional assays to test its activity, and evolutionary studies to understand its origins and adaptations. This comprehensive analysis not only enhances our understanding of the generated sequences but also opens new avenues for scientific exploration and application in synthetic biology and genetic engineering.
```
Comment (bebatut) [inline topics/statistics/tutorials/genomic-llm-sequence-generation/tutorial.md:760]:
```suggestion
After translating a synthetic DNA sequence into a protein sequence, the next crucial step is to analyze this sequence to uncover its potential biological function. This involves identifying known protein domains or motifs within the sequence that may indicate specific functions, such as binding sites or active regions. Verifying the biological plausibility of the translated sequence is essential, which can be achieved by comparing it to known protein sequences in databases like UniProt or GenBank using tools like InterProScan. Additionally, examining the conservation of the sequence across different species can provide insights into its functional importance. If the sequence shows potential for biological function, it can serve as a starting point for further research, including structural modeling to predict its three-dimensional structure, functional assays to test its activity, and evolutionary studies to understand its origins and adaptations. This comprehensive analysis not only enhances our understanding of the generated sequences but also opens new avenues for scientific exploration and application in synthetic biology and genetic engineering.
```

Comment (bebatut) [reply topics/statistics/tutorials/genomic-llm-sequence-generation/tutorial.md:760]: Better?

Comment (olisand) [reply topics/statistics/tutorials/genomic-llm-sequence-generation/tutorial.md:760]: You could leave BLAST as well: "...using tools like BLAST and InterproScan."

Comment (bebatut) [reply topics/statistics/tutorials/genomic-llm-sequence-generation/tutorial.md:760]: Updated

Comment (bgruening) [reply topics/statistics/tutorials/genomic-llm-sequence-generation/tutorial.md]: Can we please avoid useing BLAST for protein sequences, we should really use DIAMOND here. Its so much faster and does not mess with the cluster.

Comment (bebatut) [inline topics/statistics/tutorials/genomic-llm-sequence-generation/tutorial.md]:
```suggestion
After translating a synthetic DNA sequence into a protein sequence, the next crucial step is to analyze this sequence to uncover its potential biological function. This involves identifying known protein domains or motifs within the sequence that may indicate specific functions, such as binding sites or active regions. Verifying the biological plausibility of the translated sequence is essential, which can be achieved by comparing it to known protein sequences in databases like UniProt or GenBank using tools like Diamond and InterProScan. Additionally, examining the conservation of the sequence across different species can provide insights into its functional importance. If the sequence shows potential for biological function, it can serve as a starting point for further research, including structural modeling to predict its three-dimensional structure, functional assays to test its activity, and evolutionary studies to understand its origins and adaptations. This comprehensive analysis not only enhances our understanding of the generated sequences but also opens new avenues for scientific exploration and application in synthetic biology and genetic engineering.
```

**Review by olisand — APPROVED** [review]
